api/desecapi/tests/test_token_domain_policy.py

```diff
@@ -0,0 +1,395 @@
+                    _reset_policies(self.token)
+                    policy = policies[domain]
+                    setattr(policy, perm, value)
+                    policy.save()
+
+                    # Perform requests that test this permission and inspect responses
+                    for response in _perform_requests(domain.name, perm, value, **kwargs):
+                        if value:
+                            self.assertIn(response.status_code, range(200, 300))
+                        else:
+                            self.assertStatus(response, status.HTTP_403_FORBIDDEN)
+
+                    # Can't create domain
+                    data = {'name': self.random_domain_name()}
+                    response = self.client.post(self.reverse('v1:domain-list'), data, **kwargs)
+                    self.assertStatus(response, status.HTTP_403_FORBIDDEN)
+
+                    # Can't access account details
+                    response = self.client.get(self.reverse('v1:account'), **kwargs)
+                    self.assertStatus(response, status.HTTP_403_FORBIDDEN)
+
+    def test_domain_owner_consistency(self):
+        models.TokenDomainPolicy(token=self.token, domain=None).save()
+
+        domain = self.my_domains[0]
+        policy = models.TokenDomainPolicy(token=self.token, domain=domain)
+        policy.save()
+
+        domain.owner = self.other_domains[0].owner
+        with self.assertRaises(IntegrityError):
+            with transaction.atomic():  # https://stackoverflow.com/a/23326971/6867099
+                domain.save()
+
+        policy.delete()
+        domain.save()
+
+    def test_token_user_consistency(self):
+        policy = models.TokenDomainPolicy(token=self.token, domain=None)
+        policy.save()
+
+        self.token.user = self.other_domains[0].owner
+        with self.assertRaises(IntegrityError):
+            with transaction.atomic():  # https://stackoverflow.com/a/23326971/6867099
+                self.token.save()
+
+        policy.delete()
+        self.token.save()
+
+    def test_domain_owner_equals_token_user(self):
+        models.TokenDomainPolicy(token=self.token, domain=None).save()
+
+        with self.assertRaises(IntegrityError):
+            with transaction.atomic():  # https://stackoverflow.com/a/23326971/6867099
+                models.TokenDomainPolicy(token=self.token, domain=self.other_domains[0]).save()
+
+        self.token.user = self.other_domain.owner
+        with self.assertRaises(IntegrityError):
+            with transaction.atomic():  # https://stackoverflow.com/a/23326971/6867099
+                self.token.save()
+
+    def test_domain_deletion(self):
+        domains = [None] + self.my_domains[:2]
+        for domain in domains:
+            models.TokenDomainPolicy(token=self.token, domain=domain).save()
+
+        domain = domains.pop()
+        domain.delete()
+        self.assertEqual(list(map(lambda x: x.domain, self.token.tokendomainpolicy_set.all())), domains)
```

Comment:
no domains property at `tokendomainpolicy_set`?

Reply:
no: `tokendomainpolicy_set` is a set of policies, and each policy only has *one* domain. The set collection does not have a "collective attribute" to access attributes of its members (`{p.domain for p in tokendomainpolicy_set}` is not generically exposed).